### PR TITLE
cos-lat averaging

### PR DESCRIPTION
**Pull request by RondeauG**
### Pull Request Checklist:
- [x] `pre-commit` hooks are installed/active in my local clone (`$ pre-commit install`)
- [ ] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes #xyz
- [x] (If applicable) Documentation has been added / updated (for bug fixes / features)
- [x] HISTORY.rst has been updated (with summary of main changes)
  - [x] Link to issue (:issue:`number`) and pull request (:pull:`number`) has been added

### What kind of change does this PR introduce?

* Added `cos-lat` method to `aggregate.spatial_mean()`
* Deprecated `mean` and added warnings to `xesmf`
* Reworked the warning messages in `_subset_file_coverage` to be more explicit
* Fixed a bug in `compute_indicators` when less than 3 timesteps were produced.

### Does this PR introduce a breaking change?

* No, but deprecates `mean` and changes the name of `interp_coord` to `interp_centroid` (still supports `interp_coord` with a deprecation warning)

### CODE REVIEW

**Comment by RondeauG** [comment]
This is a partial fix to #94. Fixing `xesmf.SpatialAverager` is harder than expected and might require a PR in `xESMF` itself, but I didn't want to wait that long to implement cos-lat averaging.

File: xscen/aggregate.py
Comment (aulemahal) [inline xscen/aggregate.py:402]: I'm not sure I understand why you are doing this over the "Y" dimension and not over "latitude" directly. Like, why is 
```
weight = np.cos(np.deg2rad(ds.cf['latitude']))
weight = weight / weight.sum()
```
not correct?

AFAIU, this also support 2D grids.

You need `spatial_dims` for the effective `mean` call, but `spatial_dims["Y"]` is `rlat` in rotated pole, which is NOT wanted for the `cos-lat` method, no?

File: xscen/indicators.py
Comment (aulemahal) [inline xscen/indicators.py:142]: Do you think ii would be better to wrap this in a private function instead of having it twice?

**Review by aulemahal — CHANGES_REQUESTED** [review]
A few comments. I don't think the "cos-lat" think is compatible with curvilinear grids, but I'm not sure why it shouldn't be.

File: xscen/aggregate.py
Comment (mccrayc) [reply xscen/aggregate.py:402]: Yes, we should be using 'lat' rather than 'rlat' for the weights in these cases, and I can confirm that using ds.cf['latitude'] (rather than spatial_dims['Y']) does work with 2D rotated grids. I tested this with a CRCM5 monthly statistics file with spatial_dims rlat and rlon, and the following works properly: 
```
weight = np.cos(np.deg2rad(ds.cf['latitude'])) 
da_weighted = ds['tas_moy'].weighted(weight).mean([d for d in spatial_dims.values()])
```

**Review by mccrayc — CHANGES_REQUESTED** [review]
Two key points for the cos-lat averaging:
- No need to normalize the weights by their sum, unless I'm misunderstanding something xarray does this already.
- In agreement with what Pascal said, I believe cos-lat weighting should work fine with 2D curvilinear grids with a couple minor modifications (see comments).

**Comment by aulemahal** [comment]
Je me reprends, `cos-lat` as implemented is indeed not usable with curvilinear grids. The thing is, `cos(lat)` is a weight that varies in latitude and thus assumes all cells have the same area in °² (degrees squared). Which is not true for rotated pole for example.

I found this blog : https://towardsdatascience.com/the-correct-way-to-average-the-globe-92ceecd172b7
It shows a method approximating the cell area with a lat-dependent earth radius. It however uses a cell area approximation that only works for non-curvilinear grids. You could use `cf_xarray` 0.7.6 to compute 2D bounds and then cell areas (in °).

@RondeauG your code was indeed correct. But I think curvilinear grid support is essential as one of the principal use case is with MRCC5 data.

**Comment by aulemahal** [comment]
@RondeauG  I found the challenge interesting and spent a bit too much time on this, but here are 3 ways of computing cell areas for the weighted average:
https://gist.github.com/aulemahal/13c907cf2f7ccd3a84be34a7a009516c

We might not have `pygeos` already, but it's an optional dependency of `clisops`, so it might be worth adding it explicitly, it is much faster!

However, there's another issue with `cf-xarray`... Computing the 2D bounds yields strange results when the cell crosses the first lon meridian. In the MRCC5 case, the domain corner  crosses the +180 / -180 line and cell areas are super large because the corners are wrong. 

See:
image

**Comment by RondeauG** [comment]
As a reference to compare to, I implemented the cos-lat method for 2D grids using `ds.cf['latitude']` instead of `ds.cf['X']`. Here's an example for RegCM4 / NAM-22:

image

**Comment by RondeauG** [comment]
I'll look into the links provided, though I feel like this might be another method entirely? As in, we'd have `cos-lat` or `SpatialAverager` or `pygeos/ESMF/whatever`?

**Comment by RondeauG** [comment]
Also, with the changes, `get_spatial_dims(ds)` is pretty useless. I could restore it within `method=="mean"` as it was before, but I'm tempted to just get rid of `mean` entirely since it does the wrong thing to begin with.

**Comment by aulemahal** [comment]
> this might be another method entirely?
My "shapely" and "pygeos" functions are both based on the cos-lat approximation, so I would argue it's just a curvilinear-compatible version of `cos-lat`.

The ESMF version is indeed another technique because it computes the cell area using other approximations. 

Note for shapely:  the newest version (2.0) now merges the functionality of `pygeos` so there might be a way to code this idea without adding a new dependency.

**Comment by aulemahal** [comment]
I updated my gist with a function that works with shapely 2.0. It is indeed the exact same as with `pygeos` and it's as fast.

**Comment by RondeauG** [comment]
@aulemahal @mccrayc 
This might be a lack of comprehension from my part on rotated grids, but let's say that I have this region:

image

The first part of Pascal's code (using the `pygeos` version for example) produces this:
`xr.DataArray(pygeos.area(pygeos.polygons(pygeos.linearrings(ds.lon_bounds, ds.lat_bounds))), dims=ds.lon.dims, coords=ds.lon.coords).plot()`

image

`np.cos(np.deg2rad(ds.lat))` creates more or less the opposite:

image

Meaning that when you multiply both, you get:

image

...which makes no intuitive sense to me? I would have expected weights to be curved, not linear like that?

**Comment by aulemahal** [comment]
No I think this is ok! You are in a Oblique  Mercator projection, which projects the globe unto a cylinder. The projection being centered at y = 0, in the center of your plot, it makes sense to me that the cell area decreases with `abs(y)`.

https://en.wikipedia.org/wiki/Oblique_Mercator_projection

Also, I found a way to compute areas (in meters) with `pyproj` from lat and lon bounds. This was added to the gist, and gives the following:
image

File: xscen/aggregate.py
Comment (aulemahal) [inline xscen/aggregate.py:373]:
```suggestion
        if (
            ((ds.cf["longitude"].min() < -160) & (ds.cf["longitude"].max() > 160))
            or ((ds.cf["longitude"].min() < 20) & (ds.cf["longitude"].max() > 340))
        ):
```
To catch the other edge case.

File: xscen/aggregate.py
Comment (aulemahal) [inline xscen/aggregate.py:395]: I'm not sure how to phrase it, but I feel this should mention that we are also weighting by the cell area in deg².

**Review by aulemahal — APPROVED** [review]
This looks good!

Semi-related:
As we currently have no bulletproof method, the choice makes sense. However, it seems to me that there's no reason to have such an array of methods for such a simple task. I guess that when we have a simple, light area-weighted mean method, we could remove everything else? Is there a use for a collection of imperfect algos ?

This would mean :
- fix bounds generation when crossing meridians in `cf_xarray`
- generate `areacella` in m² from those bounds (using `pyproj` and `shapely`)

**Comment by RondeauG** [comment]
> Semi-related: As we currently have no bulletproof method, the choice makes sense. However, it seems to me that there's no reason to have such an array of methods for such a simple task. I guess that when we have a simple, light area-weighted mean method, we could remove everything else? Is there a use for a collection of imperfect algos ?

That's why I deprecate `mean`, since `cos-lat` is a superior version of the same thing. I could see an argument to drop `interp_centroid` in favor of always using `xesmf` too. 

However, I think that `cos-lat` and `xesmf` (SpatialAverager) fundamentaly do different things. `cos-lat` keeps the grid cells as is, while `SpatialAverager` will break pixels into smaller parts to closely fit to a given polygon.

File: xscen/aggregate.py
Comment (RondeauG) [reply xscen/aggregate.py:397]: `using a 'cos-lat' approximation of areacella (in deg2)` ?

Comment (aulemahal) [reply xscen/aggregate.py]: Agreed!

**Comment by aulemahal** [comment]
@mccrayc When you have time, a second review is needed since you "requested changes".

**Review by mccrayc — APPROVED** [review]
Looks good to me!

File: xscen/aggregate.py
Comment (RondeauG) [inline xscen/aggregate.py:459]:
```suggestion
            "A bug has been found with xesmf.SpatialAverager that appears to impact big regions. "
```

File: xscen/indicators.py
Comment (RondeauG) [inline xscen/indicators.py:168]:
```suggestion
        # Make sure that attributes have been kept for the dimensions and coordinates. Fixes a bug in xarray.
```

File: xscen/aggregate.py
Comment (aulemahal) [inline xscen/aggregate.py:348]:
```suggestion
            category=FutureWarning,
        )
    elif method == "interp_coord":
        warnings.warn(
            "xs.spatial_mean with method=='interp_coord' is deprecated. Use method=='interp_centroid' instead.",
            category=FutureWarning,
```
From python's doc:

**DeprecationWarning** : Base category for warnings about deprecated features when those warnings are intended for other Python developers (ignored by default, unless triggered by code in __main__).
**FutureWarning** : Base category for warnings about deprecated features when those warnings are intended for end users of applications that are written in Python.

Also, `stacklevel=2` lands in the `parse_config` wrapper. Without it (=1), the warning is shown was stemming from inside the function. And with 3, it stems from call of `spatial_mean` (like the workflow or the notebook cell itself).

**Pull request merged**
